include/onnxruntime/core/session/onnxruntime_c_api_ep.h

```diff
@@ -0,0 +1,185 @@
+  OrtCreateStream* create_stream;
+  const OrtDevice* default_device;
+  void* extra_param_for_create_state_func;
+  void* extra_param_for_compute_func;
+} OrtExecutionProvider;
+
+typedef struct OrtExecutionProviderFactory {
+  OrtExecutionProvider*(ORT_API_CALL* CreateExecutionProvider)(OrtExecutionProviderFactory* this_, const char* const* ep_option_keys, const char* const* ep_option_values, size_t option_size);
+} OrtExecutionProviderFactory;
+
+struct OrtGraphApi {
+const char*(ORT_API_CALL* OrtGraph_GetName)(const OrtGraphViewer*)NO_EXCEPTION ORT_ALL_ARGS_NONNULL;
+
+bool(ORT_API_CALL* OrtGraph_IsConstantInitializer)(const OrtGraphViewer* graph, const char* name, bool check_outer_scope)NO_EXCEPTION ORT_ALL_ARGS_NONNULL;
+
+size_t(ORT_API_CALL* OrtGraph_GetNodesIndexInTopologicalOrder)(const OrtGraphViewer* graph, int execution_order, _Out_ const size_t** nodes_index_in_topological_order);
+
+bool(ORT_API_CALL* OrtGraph_IsSubgraph)(const OrtGraph* graph);
+
+const OrtGraph*(ORT_API_CALL* OrtGraph_GetParentGraph)(const OrtGraph* graph);
+
+const OrtNode*(ORT_API_CALL* OrtGraph_GetParenNode)(const OrtGraphViewer* graph);
+
+const void*(ORT_API_CALL* OrtGraph_GetModelPath)(const OrtGraphViewer* graph);
+
+const OrtGraph*(ORT_API_CALL* OrtGraph_GetOrtGraph)(const OrtGraphViewer* graph_viewer);
+
+size_t(ORT_API_CALL* OrtGraph_GetInputsIncludingInitializers)(const OrtGraphViewer* graph, _Outptr_ const char*** input_names);
+
+const OrtNode*(ORT_API_CALL* OrtGraph_GetOrtNode)(const OrtGraphViewer* graph, size_t node_index);
+
+size_t(ORT_API_CALL* OrtGraph_GetNodesConsumingInput)(const OrtGraphViewer* graph, const char* input_name, _Outptr_ const OrtNode*** consumers); // TODO(leca): ValueConsumers::comprehensive ?
+
+const OrtNode*(ORT_API_CALL* OrtGraph_GetNodeProducingOutput)(const OrtGraphViewer* graph, const char* output_name);
+
+int(ORT_API_CALL* OrtGraph_NumberOfNodes)(const OrtGraphViewer*)NO_EXCEPTION ORT_ALL_ARGS_NONNULL;
+
+int(ORT_API_CALL* OrtGraph_MaxNodeIndex)(const OrtGraphViewer* graph);
+
+size_t(ORT_API_CALL* OrtGraph_GetOutputSize)(const OrtGraphViewer*)NO_EXCEPTION ORT_ALL_ARGS_NONNULL;
+
+const char*(ORT_API_CALL* OrtGraph_GetIthOutputName)(const OrtGraphViewer*, size_t i)NO_EXCEPTION ORT_ALL_ARGS_NONNULL;
+
+int32_t(ORT_API_CALL* OrtGraph_GetIthOutputElemType)(const OrtGraphViewer*, size_t i)NO_EXCEPTION ORT_ALL_ARGS_NONNULL;
+
+bool(ORT_API_CALL* OrtGraph_GetInitializerTensor)(const OrtGraphViewer* graph, const char* initializer_name, _Outptr_ OrtTensorRef**);
+
+bool(ORT_API_CALL* OrtGraph_GetValueInfo)(const OrtGraphViewer* graph, const char* name, _Outptr_ OrtValueInfoRef**);
+
+size_t(ORT_API_CALL* OrtGraph_SerializeToArray)(const OrtGraphViewer*, _Out_ void** data)NO_EXCEPTION;  // TODO(leca): review and discuss
+
+ORT_API2_STATUS(OrtGraph_GetSubGraph, const OrtGraphViewer* graph, const int node_num, const size_t* node_indices, _Outptr_ const OrtGraphViewer** subgraph); // TODO(yang): review and discuss
```

Comment:
Could we add an additional graph api or feature to check whether the input of the subgraph has shape info?

For TRT EP, TRT requires onnx model to have input shape provided, so that's why TRT EP checks the input shape of the subgraph, if not it suggests user to run symbolic_shape_infer.py first.

Reply:
Ideally this is via more basic functions. An API to get the graph input names, and an API to get the NodeArg entry (in some more easily digested format) for that value.

The checking whether it has shape info or not seem EP specific and not something the general API needs to directly support.

Reply:
We have that already: OrtGraph_GetInputsIncludingInitializers (naming is subject to change per Scott's comment) and OrtGraph_GetValueInfo